Add loading feedback for email change in user management

MorrisJobke opened the pull request:
- show loading indicator while the request is running
- change value after the request succeed - not before and
  undo maybe wrong changes

The loading indicator is to the right of the input field

![bildschirmfoto am 2015-04-20 um 14 36 14](https://cloud.githubusercontent.com/assets/245432/7230064/b4d93d24-e76a-11e4-8948-17d265924e7a.png)

cc @LukasReschke @jancborchardt

MorrisJobke committed:
Add loading feedback for email change in user management

* show loading indicator while the request is running
* change value after the request succeed - not before and
  undo maybe wrong changes
* move markup to HTML template
* show notification temporarily

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6cfb80c8-cc9d-483f-afa6-2013b3f42158) was created.

LukasReschke commented:
Works :+1:

DeepDiver1975 commented on settings/js/users/users.js:
```diff
@@ -733,30 +733,42 @@ $(document).ready(function () {
+					var mailAddress = $input.val();
+
+					if (mailAddress.length > 0) {
+						var $loading = $('<div>')
```
hmmm - what about adding the div by default (hidden for sure) and only show it in here?

MorrisJobke replied:
I tried it, but then the hidden logic f**ks all up. I want to add a element with `display: inline-block`. `.hidden` uses also `display` with the value `hidden`. I then need to toggle the `display`value inside of the JS, which is also ugly and with cause more headaches, because then you have the markup in HTML, some rules in CSS and the CSS rules after an interaction in the JS part. Therefore I thought it would be easier to maintain it in JS only.

DeepDiver1975 replied:
brrrr - soooo ugly .... :see_some_evil: :wink: 

@PVince81 @jancborchardt any ideas on this - THX

DeepDiver1975 commented:
@MorrisJobke does this fix any existing issue? please link - tx

MorrisJobke commented:
@DeepDiver1975 No, but currently it works like this:

### Old:

* Click edit -> input appears
* change
* hit enter
* it will remove the input and place the content of the input as text 
* send request
* success: all is fine
* failure: it will show a notification and switch back the value

### New:

* Click edit -> input appears
* change
* hit enter
* send request - show loading indicator
* success: all is fine - switch the input to plain text and show the new value
* failure: keep the input with the wrong value and show a notification

DeepDiver1975 commented:
@owncloud-bot retest this please

MorrisJobke commented:
> Refer to this link for build results (access rights to CI server needed):
> https://ci.owncloud.org//job/pull-request-analyser-ng-simple/11678/
> Build result: FAILURE

@owncloud-bot Are you drunk?! You didn't finished yet and all tests passed until now!

nickvergessen commented:
Could also use `OC.msg.startSaving()` and `OC.msg.finishedSaving()` like other stuff on admin/personal page do, to be aligned with one design.

jancborchardt commented:
@MorrisJobke minor detail: Can you put the loading spinner inside the input field, on the right of it? That will make it clear that it belongs to that field and it’ll also prevent any potential layout shifting.

MorrisJobke commented:
> @MorrisJobke minor detail: Can you put the loading spinner inside the input field, on the right of it? That will make it clear that it belongs to that field and it’ll also prevent any potential layout shifting.

Feel free to do it. Feel also free to switch to a separate HTML layout and only class togglings in JS

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6cfb80c8-cc9d-483f-afa6-2013b3f42158) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/be37116c-7504-4221-8208-d2736adb100a) was created.

MorrisJobke commented:
Okay ... so I have no idea, why my commit with only changing the layout so heavily breaks the logic 039216c6408422c96a340025ddb3c2fecb5fbeea

Now the behaviour feels very strange. I need some help ... or just days without looking at this and then fix it maybe later.

jancborchardt commented:
Looks fine to me. :+1:

MorrisJobke commented:
This currently doesn't work as it is now :-1: We need to work on this.

MorrisJobke commented:
Currently there are following two problems with the current implementation and I have no idea how to fix it:

In Chromium:

* enter a wrong email -> it gets reset, the input field vanishes and the error message is shown (but input field should stay with the wrong value, to easiely be able to fix the error) - somehow the blur event gets triggered when pressing Enter, but `event.preventDefault()` doesn't help here to not fire the blur event.

In Firefox:

* enter a correct email -> the input field stays with the new email, but it should go away and show the updated email as text (not within the input field) - somehow the blur event isn't triggered in the success case. Once you debug this step by step it works. I guess this is because the mouse pointer is then outside of the window and another blur event gets triggered -.-


### How it should work

* type in a email address
* press enter
* get feedback (loading spinner)
* success: remove input and show the updated email address
* error: show error message and still show the input field with the wrong value to be able to update it again

MorrisJobke commented:
maybe @PVince81 can help here as our JS guru :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/541f25b7-5856-4531-a7ac-b9b67ca4cf9c) was created.

MorrisJobke commented:
@jancborchardt We need to talk about this during conf

DeepDiver1975 commented:
@MorrisJobke @jancborchardt close? finish? THX

MorrisJobke commented:
> @MorrisJobke @jancborchardt close? finish? THX

I would like to have a interaction design first. What should happen when which of the following actions are done:

* press enter
* tab out of/un-focus the input field
* navigate away: close tab, browse to other app, search for another user, delete another user, delete the same user (especially regarding undo this action then) ...

And how to cancel the update - when I removed accidentally the content and just want to undo it?

@owncloud/designers

jancborchardt commented:
@MorrisJobke can you add an .icon-confirm button to the right of the field just like I did in the log in screen and the Bookmarks app? See https://github.com/owncloud/bookmarks/pull/170

On top of that, to answer your questions: only pressing enter and pressing that button should apply the change. Tabbing out or anything else should not save.

MorrisJobke commented:
> On top of that, to answer your questions: only pressing enter and pressing that button should apply the change. Tabbing out or anything else should not save.

Okay. Thanks.

jancborchardt commented:
@MorrisJobke any update on this? Or is https://github.com/owncloud/core/issues/16360 already fixed?

MorrisJobke commented:
> @MorrisJobke any update on this? Or is #16360 already fixed?

I will check this soon.

MorrisJobke commented:
> @MorrisJobke can you add an .icon-confirm button to the right of the field just like I did in the log in screen and the Bookmarks app? See owncloud/bookmarks#170

I just rebased this PR, fixed the different state changes (blur vs enter), handled errors properly and added the spinner once the request is send. I would like to get this in as it is. The confirm icon is not yet included, but I would do that as a separate PR which then also would fix all other inputs too.

@PVince81 @owncloud/javascript @blizzz Please review this.

MorrisJobke commented:
I tested this in Firefox, Safari, IE11 and Chrome. All of them now work the same.

MorrisJobke commented:
The spinner is BTW inside the input field ;)

MorrisJobke commented:
> @PVince81 @owncloud/javascript @blizzz Please review this.

and maybe also @nickvergessen @owncloud/designers could help reviewing this :)

skjnldsv commented:
Good here! :+1:

LukasReschke commented:
https://github.com/owncloud/core/pull/15750#issuecomment-94442419 still applies :+1:
